### PR TITLE
Improve is_allowed_to_start

### PR DESCRIPTION
**Pull request by predragnikolic**
As it doesn't allow the server to initialize if the tailwind.config.js is in a sub directory.

### CODE REVIEW

**Comment by rchl** [comment]
But it triggers errors like:

```
Traceback (most recent call last):
  File "/me/Sublime Text/Packages/LSP/plugin/session_view.py", line 172, in on_capability_added_async
    self._register_auto_complete_triggers(registration_id, trigger_chars or [])
  File "/me/Sublime Text/Packages/LSP/plugin/session_view.py", line 96, in _register_auto_complete_triggers
    self._apply_auto_complete_triggers(self.view.settings(), trigger_chars, registration_id)
  File "/me/Sublime Text/Packages/LSP/plugin/session_view.py", line 130, in _apply_auto_complete_triggers
    trigger["characters"] = "".join(trigger_chars)
TypeError: sequence item 8: expected str instance, NoneType found
```

when there is no config file.

Have you checked what the server actually supports and whether we should maybe replicate that? Where does it search for the config if not only in the root? Is it relative to the opened file?

**Comment by predragnikolic** [comment]
You are right.

We should probably replicate what the vs code extension does - https://github.com/tailwindlabs/tailwindcss-intellisense/blob/766a5d533dcb68640ce6b3270488f6701dd1173d/packages/vscode-tailwindcss/src/extension.ts#L378

**Comment by predragnikolic** [comment]
I could use wcmatch dependency to find the config file in the sub folder https://facelessuser.github.io/wcmatch/glob/

**Comment by rwols** [comment]
The glob is so simple, why not use `os.walk` and check for each file if it’s a filename of interest? No need for dependencies then.

**Comment by predragnikolic** [comment]
Thanks for suggesting `os.walk`

File: plugin.py [hunk 1]
```diff
@@ -29,9 +30,26 @@ def is_allowed_to_start(
...
+            return None # config found, return None to start the session
+        return "No tailwind configuration file present in the workspace folder."
+
+
+def find_file_in_workspace(file_pattern: str, root_folder: str, folder_exclude_patterns: Optional[List[str]] = None) -> Optional[str]:
+    for path, directories, files in os.walk(root_folder):
+        if folder_exclude_patterns:
+            for folder_excldue_pattern in folder_exclude_patterns:
+                # skip ignored folders
+                directories[:] = [d for d in directories if not re.search(folder_excldue_pattern, d)]
```
Comment (rwols) [inline plugin.py:50]: This is a bit odd. the folder_exclude_pattern is a glob, but we're using it as if it's a regex.

Comment (rchl) [inline plugin.py:50]: The exclude pattern is not a regexp so using `re.search` is not correct. If you want to handle wildcards in the pattern then you'd need something more sophisticated. If not handling wildcards is OK then just checking if exists in the list would maybe suffice.

File: plugin.py [hunk 2]
```diff
@@ -29,9 +30,26 @@ def is_allowed_to_start(
...
-        tailwind_config_file_path = os.path.join(path, 'tailwind.config.js')
-        if not os.path.exists(tailwind_config_file_path):
-            return "No tailwind.config.js present in {}".format(path)
-        return None
 
+        # Config pattern is found here:
+        # https://github.com/tailwindlabs/tailwindcss-intellisense/blob/766a5d533dcb68640ce6b3270488f6701dd1173d/packages/vscode-tailwindcss/src/extension.ts#L40
+        config_file_pattern = '(tailwind|tailwind.config).(js|cjs)'
+        folder_exclude_patterns = sublime.load_settings('Preferences.sublime-settings').get("folder_exclude_patterns", []) # type: List[str]
+        folder_exclude_patterns.append('node_modules') # definitely exclude node_modules
```
Comment (rchl) [inline plugin.py:38]: Converting to `Set` would result in faster lookup and avoid potential duplicates.

File: plugin.py [hunk 3]
```diff
@@ -29,9 +30,26 @@ def is_allowed_to_start(
...
             return "Requires a folder to start."
-        path = workspace_folders[0].path
-        tailwind_config_file_path = os.path.join(path, 'tailwind.config.js')
-        if not os.path.exists(tailwind_config_file_path):
-            return "No tailwind.config.js present in {}".format(path)
-        return None
 
+        # Config pattern is found here:
+        # https://github.com/tailwindlabs/tailwindcss-intellisense/blob/766a5d533dcb68640ce6b3270488f6701dd1173d/packages/vscode-tailwindcss/src/extension.ts#L40
+        config_file_pattern = '(tailwind|tailwind.config).(js|cjs)'
```
Comment (rchl) [inline plugin.py:36]: The dots need to be escaped.
And there could be a `$` at the end.

**Review by rchl — CHANGES_REQUESTED** [review]

File: plugin.py [hunk 4]
```diff
@@ -33,21 +34,20 @@ def is_allowed_to_start(
...
 
 
-def find_file_in_workspace(file_pattern: str, root_folder: str, folder_exclude_patterns: Optional[List[str]] = None) -> Optional[str]:
+def find_file_in_workspace(file_pattern: str, root_folder: str, folder_exclude_patterns: Set[str] = None) -> Optional[str]:
     for path, directories, files in os.walk(root_folder):
-        if folder_exclude_patterns:
-            for folder_excldue_pattern in folder_exclude_patterns:
-                # skip ignored folders
-                directories[:] = [d for d in directories if not re.search(folder_excldue_pattern, d)]
+        for folder_excldue_pattern in folder_exclude_patterns or []:
```
Comment (rchl) [inline plugin.py]:
```suggestion
        for folder_exclude_pattern in folder_exclude_patterns or []:
```

**Comment by predragnikolic** [comment]
Sublime supports https://www.sublimetext.com/docs/file_patterns.html just `*` and `?` file patterns which luckily fnmatch supports :)

File: plugin.py [hunk 5]
```diff
@@ -29,9 +31,25 @@ def is_allowed_to_start(
...
+
+
+def find_file_in_workspace(file_pattern: str, root_folder: str, folder_exclude_patterns: Set[str] = None) -> Optional[str]:
+    for path, directories, files in os.walk(root_folder):
+        for folder_exclude_pattern in folder_exclude_patterns or []:
+            # skip ignored folders
+            directories[:] = [d for d in directories if not fnmatch.fnmatch(d, folder_exclude_pattern)]
+
+        for file in files:
+            if re.search(file_pattern, file):
```
Comment (rchl) [inline plugin.py:53]: `re.match` should be a better fit in this case.

**Review by rchl — APPROVED** [review]

**Comment by rchl** [comment]
> Sublime supports [sublimetext.com/docs/file_patterns.html](https://www.sublimetext.com/docs/file_patterns.html) just `*` and `?` file patterns which luckily fnmatch supports :)

Those are not entirely equivalent because handling of forward-slashes is different. In ST `*` doesn't match across those but it does in `fnmatch` (mainly because `fnmatch` is supposed to be used to match file names only).

**Comment by predragnikolic** [comment]
> Those are not entirely equivalent because handling of forward-slashes is different. In ST * doesn't match across those but it does in fnmatch (mainly because fnmatch is supposed to be used to match file names only).

If the fnmatch causes some unexpected issues with folder_exclude_patterns. I will switch from patters to regular strings
```
directories[:] = [d for d in directories if d not in ['node_modules', '.git']] 
```

**Pull request merged**